pkg/virt-api/rest/subresource.go

```diff
@@ -464,9 +489,62 @@ func (app *SubresourceAPIApp) RestartVMRequestHandler(request *restful.Request,
 		return
 	}
 
+	// Only force restart with GracePeriodSeconds=0 is supported for now
+	// Here we are deleting the Pod because CRDs don't support gracePeriodSeconds at the moment
+	if bodyStruct.GracePeriodSeconds != nil {
+		if *bodyStruct.GracePeriodSeconds == 0 {
+			vmiPodname, err := app.findPod(namespace, vmi)
+			if err != nil {
+				response.WriteError(http.StatusForbidden, err)
+				return
+			}
+			// set termincationGracePeriod and delete the VMI pod to trigger a forced restart
+			err = app.virtCli.CoreV1().Pods(namespace).Delete(vmiPodname, &k8smetav1.DeleteOptions{GracePeriodSeconds: bodyStruct.GracePeriodSeconds})
+			if err != nil {
+				if errors.IsNotFound(err) {
+					// ignoring this because pod might be deleted by VM status change
+				} else {
+					response.WriteError(http.StatusForbidden, err)
+					return
+				}
+			}
+		}
+	}
+
 	response.WriteHeader(http.StatusAccepted)
 }
 
+func (app *SubresourceAPIApp) findPod(namespace string, vmi *v1.VirtualMachineInstance) (string, error) {
+	fieldSelector := fields.ParseSelectorOrDie("status.phase==" + string(v12.PodRunning))
+	labelSelector, err := labels.Parse(fmt.Sprintf(v1.AppLabel + "=virt-launcher," + v1.CreatedByLabel + "=" + string(vmi.UID)))
+	if err != nil {
+		return "", err
+	}
+	selector := k8smetav1.ListOptions{FieldSelector: fieldSelector.String(), LabelSelector: labelSelector.String()}
+	podList, err := app.virtCli.CoreV1().Pods(namespace).List(selector)
+	if err != nil {
+		return "", err
+	}
+	if len(podList.Items) == 0 {
+		return "", goerror.New("connection failed. No VirtualMachineInstance pod is running")
```

Comment:
I guess you need `return "", nil` here too.

Reply:
Changed